### ISSUE TITLE
LHError: NO_FCP

### ISSUE THREAD

**Issue by mbalabash** [aslushnikov/tracium#6]
Hello

I have next html file:
```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Estimo Template</title>
  </head>
  <body>
    <script src="/Users/mbalabash/Documents/projects/estimo/example/lib/2kb.js"></script>
  </body>
</html>
```

And the next js code with puppeteer to capture chrome trace:
```js
async function createChromeTrace(urlToHtmlFile, pathToTraceFile, browserOptions) {
  const { _headless: headless, _timeout: timeout } = browserOptions

  const browser = await puppeteer.launch({ headless })
  const page = await browser.newPage()

  await page.tracing.start({ path: pathToTraceFile })
  try {
    await page.goto(urlToHtmlFile, { timeout })
  } catch (err) {
    handleSessionError(err, browser)
  }
  await page.tracing.stop()

  await browser.close()
  return pathToTraceFile
}
```

And when i run it, i got an error: 
```
LHError: NO_FCP
    at Function.compute (/Users/mbalabash/Documents/projects/estimo/node_modules/tracium/lib/trace-of-tab.js:101:38)
    at Function.compute (/Users/mbalabash/Documents/projects/estimo/node_modules/tracium/lib/main-thread-tasks.js:271:55)
    at Object.computeMainThreadTasks (/Users/mbalabash/Documents/projects/estimo/node_modules/tracium/index.js:13:36)
```

In trace file, sections with FP / FCP / FMP really not exists. 
I checked it few times. 

But i can't reproduce this error with remote lib: 
```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Estimo Template</title>
  </head>
  <body>
    <script src="https://unpkg.com/react-dom@16/umd/react-dom.production.min.js"></script>
  </body>
</html>
```
And with next workaround it also works fine:
```js
async function createChromeTrace(urlToHtmlFile, pathToTraceFile, browserOptions) {
  const { _headless: headless, _timeout: timeout } = browserOptions

  const browser = await puppeteer.launch({ headless })
  const page = await browser.newPage()

  await page.tracing.start({ path: pathToTraceFile })
  try {
    await page.goto(urlToHtmlFile, { timeout })
    await page.waitFor(100) // <--
  } catch (err) {
    handleSessionError(err, browser)
  }
  await page.tracing.stop()

  await browser.close()
  return pathToTraceFile
}
```

**Questions**
1) Do you know why it happens?
2) Can you suggest more predictable solution or workaround?

**Comment by patrickhulce** [comment]
There are lots of subtle issues with `file://` URLs and the core of this library has assumptions that it will never be run on `file://` URLs, so YMMV there. If you can host on a local server, that would be my best advice.

If you can replicate this behavior while on `localhost`, we'd love to hear about it so we can fix :)

**Comment by aslushnikov** [comment]
@mbalabash: your example is somewhat of a corner case. If the page were to have any text or something, it would work.

@patrickhulce I can easily repro on a localhost.

I'm not even sure why it works after the `wait(100)` (and it doesn't for me). It looks like Chromium marks first contentful paint when there's [some text rendered](https://cs.chromium.org/chromium/src/third_party/blink/renderer/core/paint/paint_timing.cc?sq=package:chromium&g=0&l=135). 

My repro:

### `a.html` (statically served on `localhost:1234`)

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Estimo Template</title>
  </head>
  <body>
  </body>
</html>
```

### `a.js`
```js
const puppeteer = require('puppeteer');
const Tracium = require('tracium');

(async () => {
  const browser = await puppeteer.launch({headless: false});
  const page = await browser.newPage();
  await page.tracing.start({});
  await page.goto('http://localhost:1234/a.html');
  const buffer = await page.tracing.stop();
  await browser.close();

  const json = JSON.parse(buffer.toString('utf8'));
  const tasks = Tracium.computeMainThreadTasks(json);
  console.log(tasks.length);
})();

```

**Comment by mbalabash** [comment]
@aslushnikov I still have same error with next html:

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Estimo Template</title>
  </head>
  <body>
    <h1>Estimo</h1>
    <script src="/Users/mbalabash/Documents/projects/estimo/example/lib/2kb.js"></script>
    <h2>Estimo</h2>
  </body>
</html>
```

**Comment by patrickhulce** [comment]
@aslushnikov yes but that page on localhost has no content, so it's very expected :) I was under the assumption the script added some content to the page, but maybe I misunderstood.

@mbalabash do you see NO_FCP with that HTML on a localhost page?

I suppose it's a meta question of if `tracium` should support the cases where there's NO_FCP which I suspect the answer is yes. It's totally valid to want to parse main thread activity it will just be far more difficult in cases where Chrome fails to identify the main thread for us (started happening much more frequently in m74).

We'll have to start teasing apart all of the assumptions Lighthouse makes when parsing the trace to support those other use cases.

**Comment by aslushnikov** [comment]
@patrickhulce wdyt about not throwing an error in case there's no FCP? Lighthouse relies on FCP timing later on, but it doesn't matter much for general-case trace processing

**Comment by mbalabash** [comment]
@patrickhulce No, last test still was as `file://`.
And nope, script doesn't add some content. Just javascript.

**UPD** 
Yes. I can also reproduce error on `localhost`.

New strange detail. 
I can reproduce it only in `headless: true` mode.
It works fine in `headless: false` mode.

**Comment by patrickhulce** [comment]
Yeah SGTM @aslushnikov that's what I was trying to mean by

> We'll have to start teasing apart all of the assumptions Lighthouse makes when parsing the trace to support those other use cases.

:)

**Comment by mbalabash** [comment]
Awesome!
Thank you, guys

**Comment by mbalabash** [comment]
Works perfectly from latest `master`.
Can you publish new version to npm?

**Comment by aslushnikov** [comment]
> Can you publish new version to npm?

@mbalabash done! Published as v0.2.0

**Comment by mbalabash** [comment]
@aslushnikov Thanks!
